api/app/outcome/api.py

```diff
@@ -0,0 +1,152 @@
+import traceback
+from flask_restful import reqparse, fields, marshal_with
+import flask_restful as restful
+from flask import g, request
+from sqlalchemy.exc import SQLAlchemyError
+
+from app.outcome.models import Outcome, Status
+from app.outcome.repository import OutcomeRepository as outcome_repository
+from app.events.repository import EventRepository as event_repository
+from app.users.repository import UserRepository as user_repository
+from app.utils.emailer import send_mail
+from app.email_template.repository import EmailRepository as email_repository
+
+from app.utils.auth import auth_required, event_admin_required
+from app import LOGGER
+from app import db
+from app.utils import errors
+from app.utils import misc
+
+
+def _extract_status(outcome):
+    if not isinstance(outcome, Outcome):
+        return None
+    return outcome.status.name
+
+outcome_fields = {
+    'id': fields.Integer,
+    'status': fields.String(attribute=_extract_status),
+    'timestamp': fields.DateTime(dt_format='iso8601'),
+}
+
+user_fields = {
+    'id': fields.Integer,
+    'email': fields.String,
+    'firstname': fields.String,
+    'lastname': fields.String,
+    'user_title': fields.String
+}
+
+outcome_list_fields = {
+    'id': fields.Integer,
+    'status': fields.String(attribute=_extract_status),
+    'timestamp': fields.DateTime(dt_format='iso8601'),
+    'user': fields.Nested(user_fields),
+    'updated_by_user': fields.Nested(user_fields)
+}
+
+
+class OutcomeAPI(restful.Resource):
+    @auth_required
+    @marshal_with(outcome_fields)
+    def get(self):
+        req_parser = reqparse.RequestParser()
+        req_parser.add_argument('event_id', type=int, required=True)
+        args = req_parser.parse_args()
+
+        user_id = g.current_user['id']
+        event_id = args['event_id']
+
+        try:
+            outcome = outcome_repository.get_latest_by_user_for_event(user_id, event_id)
+            if not outcome:
+                return errors.OUTCOME_NOT_FOUND
+            
+            return outcome
+
+        except SQLAlchemyError as e:
+            LOGGER.error("Database error encountered: {}".format(e))            
+            return errors.DB_NOT_AVAILABLE
+        except: 
+            LOGGER.error("Encountered unknown error: {}".format(traceback.format_exc()))
+            return errors.DB_NOT_AVAILABLE
```

Comment:
So I was wondering if we should spend some time investigating how to globally handle errors so then we don't need to have try/except blocks in every API call. 

I did briefly look into it now and found this in the [docs](https://flask-restful.readthedocs.io/en/latest/extending.html#custom-error-handlers) but not entirely sure how the logging hooks into flask and ideally I'd like the custom error message to come from a custom exception class (i.e. we'll define a `OutcomeNotFoundError` class for example ) and if it's raised the error code and message come from within the class instead of having to be manually added to the dictionary like in the docs or returned separately like here. Not sure if something like that exists in Python/Flask. 

I know I've done something similar with a C# WebApi before. It really does clean up get/post methods etc, centralises the error handling logic, and then we can have custom error handling for application errors (say we define an `ApplicationError` from which `OutNotFoundError` inherits from) which is then globally caught, or the db errors like `SQLAlchemyError`, or any other kind of error that we want to handle separately. What do you think?

Reply:
100% agree! We should definitely do this. Most of our APIs have repeated error handling logic which is a definite code smell!